File: finding-seams/csharp/FindingSeams/ScheduleVisualising/ScheduleToText.cs
using System.Globalization;
using System.Text.Json;

namespace FindingSeams.ScheduleVisualising;

public class ScheduleToText
{
    private static readonly DateTimeOffset ConferenceStart =
        DateTimeOffset.Parse("2025-11-03T00:00:00Z", CultureInfo.InvariantCulture);

    private const int ConferenceDurationDays = 3;
    private static readonly long MsPerDay = 1000L * 60 * 60 * 24;

    private static readonly Dictionary<string, int> TimezoneOffsets = new()
    {
        ["UTC"] = 0,
        ["Europe/Helsinki"] = 120,
        ["Europe/Stockholm"] = 60,
        ["Europe/London"] = 0,
        ["America/New_York"] = -300,
        ["America/Chicago"] = -360,
        ["America/Los_Angeles"] = -480,
    };

    private static readonly HashSet<string> SupportedLocales =
        new(StringComparer.Ordinal) { "en-US", "en-GB", "fi-FI", "sv-SE", "de-DE" };

    private readonly ScheduleFormatter _formatter = new();

    public async Task<string> ConvertToText(object dateInput, ConvertToTextOptions? options = null)
    {
        options ??= new ConvertToTextOptions();

        var requested = CoerceDate(dateInput);
        if (requested is null)
            return "Invalid date provided.\n";

        var dayIndex = (int)((requested.Value.ToUnixTimeMilliseconds() - ConferenceStart.ToUnixTimeMilliseconds()) /
                             MsPerDay);
        if (dayIndex < 0 || dayIndex >= ConferenceDurationDays)
        {
            var human = requested.Value.ToString("R", CultureInfo.InvariantCulture);
            return $"No schedule available for {human}.\n";
        }

        var conferenceDay = dayIndex + 1;
        var dateString = ToDateString(requested.Value);

        var tzName = options.Timezone;
        var timezoneOffset = TimezoneOffsets.GetValueOrDefault(tzName, 0);
        var locale = options.Locale is { } loc && SupportedLocales.Contains(loc) ? loc : "en-US";
        var track = options.Track;
        var includeBreaks = options.IncludeBreaks is not false;
        var allowedTypes = options.Types ?? new[] { "talk", "workshop", "break" };

        var path = Path.Combine(AppContext.BaseDirectory, "Resources", "schedule", $"schedule-{dateString}.json");
        if (!File.Exists(path))
            return "Invalid date provided.\n";

        var raw = await File.ReadAllTextAsync(path);
        using var doc = JsonDocument.Parse(raw);
        var root = doc.RootElement;
        var sessions = new List<ScheduleSession>();
        if (root.TryGetProperty("sessions", out var arr))
        {
            foreach (var el in arr.EnumerateArray())
                sessions.Add(ParseSession(el));
        }

        if (track != "all")
            sessions = sessions.Where(s => s.Track == track || s.Track == "all").ToList();

        if (!includeBreaks)
            sessions = sessions.Where(s => s.Type != "break").ToList();

        sessions = sessions.Where(s => s.Type is { } t && allowedTypes.Contains(t)).ToList();

        var formattingOptions = new ScheduleFormatOptions(
            Locale: locale,
            TimezoneOffset: timezoneOffset,
            ConferenceDay: conferenceDay,
            Timezone: tzName);

        return _formatter.Format(new ScheduleObject(sessions), formattingOptions);
    }

    private static DateTimeOffset? CoerceDate(object dateInput) =>
        dateInput switch
        {
            DateTimeOffset dto => dto,
            DateTime dt => new DateTimeOffset(dt),
            long l => DateTimeOffset.FromUnixTimeMilliseconds(l),
            int i => DateTimeOffset.FromUnixTimeMilliseconds(i),
            uint ui => DateTimeOffset.FromUnixTimeMilliseconds(ui),
            string s when s.All(char.IsDigit) => DateTimeOffset.FromUnixTimeMilliseconds(long.Parse(s)),
            string s => DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var p)
                ? p
                : null,
            _ => null,
        };

    private static string ToDateString(DateTimeOffset instant)
    {
        var u = instant.ToOffset(TimeSpan.Zero);
        return $"{u.Year:D4}-{u.Month:D2}-{u.Day:D2}";
    }

    private static ScheduleSession ParseSession(JsonElement o)
    {
        string? Str(string name) => o.TryGetProperty(name, out var p) && p.ValueKind != JsonValueKind.Null
            ? p.GetString()
            : null;

        int? Len = o.TryGetProperty("length", out var lp) && lp.ValueKind != JsonValueKind.Null
            ? lp.GetInt32()
            : null;

        return new ScheduleSession(Str("start"), Len, Str("title"), Str("speaker"), Str("room"), Str("track"),
            Str("type"));
    }
}

File: finding-seams/kotlin/src/main/kotlin/findingseams/schedulevisualising/ScheduleToText.kt
package findingseams.schedulevisualising

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

private val CONFERENCE_START = Instant.parse("2025-11-03T00:00:00Z")
private const val CONFERENCE_DURATION_DAYS = 3
private val MS_PER_DAY = 1000L * 60 * 60 * 24

private val TIMEZONE_OFFSETS = mapOf(
    "UTC" to 0,
    "Europe/Helsinki" to 120,
    "Europe/Stockholm" to 60,
    "Europe/London" to 0,
    "America/New_York" to -300,
    "America/Chicago" to -360,
    "America/Los_Angeles" to -480,
)

private val SUPPORTED_LOCALES = setOf("en-US", "en-GB", "fi-FI", "sv-SE", "de-DE")

data class ConvertToTextOptions(
    val timezone: String = "UTC",
    val locale: String? = null,
    val track: String = "all",
    val includeBreaks: Boolean? = null,
    val types: List<String>? = null,
)

class ScheduleToText(
    private val formatter: ScheduleFormatter = ScheduleFormatter(),
) {
    suspend fun convertToText(dateInput: Any, options: ConvertToTextOptions = ConvertToTextOptions()): String =
        withContext(Dispatchers.IO) {
            val requested = coerceDate(dateInput) ?: return@withContext "Invalid date provided.\n"

            val dayIndex = ((requested.toEpochMilli() - CONFERENCE_START.toEpochMilli()) / MS_PER_DAY).toInt()
            if (dayIndex < 0 || dayIndex >= CONFERENCE_DURATION_DAYS) {
                val human = DateTimeFormatter.RFC_1123_DATE_TIME.withZone(ZoneOffset.UTC).format(requested)
                return@withContext "No schedule available for $human.\n"
            }

            val conferenceDay = dayIndex + 1
            val dateString = toDateString(requested)

            val tzName = options.timezone
            val timezoneOffset = TIMEZONE_OFFSETS[tzName] ?: 0
            val locale = options.locale?.takeIf { it in SUPPORTED_LOCALES } ?: "en-US"
            val track = options.track
            val includeBreaks = options.includeBreaks != false
            val allowedTypes = options.types ?: listOf("talk", "workshop", "break")

            val resourcePath = "/schedule/schedule-$dateString.json"
            val stream = ScheduleToText::class.java.getResourceAsStream(resourcePath)
                ?: return@withContext "Invalid date provided.\n"
            val raw = InputStreamReader(stream, StandardCharsets.UTF_8).use { it.readText() }
            val root = JsonParser.parseString(raw).asJsonObject
            val sessionsArray = root.getAsJsonArray("sessions") ?: JsonArray()

            var sessions = sessionsArray.mapNotNull { parseSession(it.asJsonObject) }.toMutableList()
            if (track != "all") {
                sessions = sessions.filter { it.track == track || it.track == "all" }.toMutableList()
            }
            if (!includeBreaks) {
                sessions = sessions.filter { it.type != "break" }.toMutableList()
            }
            sessions = sessions.filter { it.type != null && it.type in allowedTypes }.toMutableList()

            val formattingOptions = ScheduleFormatOptions(
                locale = locale,
                timezoneOffset = timezoneOffset,
                conferenceDay = conferenceDay,
                timezone = tzName,
            )
            formatter.format(ScheduleObject(sessions), formattingOptions)
        }

    private fun coerceDate(dateInput: Any): Instant? =
        when (dateInput) {
            is Instant -> dateInput
            is Number -> Instant.ofEpochMilli(dateInput.toLong())
            is String -> {
                if (dateInput.all { it.isDigit() }) {
                    Instant.ofEpochMilli(dateInput.toLong())
                } else {
                    try {
                        Instant.parse(dateInput)
                    } catch (_: Exception) {
                        try {
                            java.time.LocalDate.parse(dateInput).atStartOfDay(ZoneOffset.UTC).toInstant()
                        } catch (_: Exception) {
                            null
                        }
                    }
                }
            }
            else -> null
        }

    private fun toDateString(instant: Instant): String {
        val d = instant.atZone(ZoneOffset.UTC).toLocalDate()
        return "%04d-%02d-%02d".format(d.year, d.monthValue, d.dayOfMonth)
    }

    private fun parseSession(o: JsonObject): ScheduleSession {
        fun str(key: String): String? =
            if (!o.has(key) || o.get(key).isJsonNull) null else o.get(key).asString

        fun intOrNull(key: String): Int? =
            if (!o.has(key) || o.get(key).isJsonNull) null else o.get(key).asInt

        return ScheduleSession(
            start = str("start"),
            length = intOrNull("length"),
            title = str("title"),
            speaker = str("speaker"),
            room = str("room"),
            track = str("track"),
            type = str("type"),
        )
    }
}
